File: solutions/4095-number-of-balanced-integers-in-a-range/number-of-balanced-integers-in-a-range.java
class Solution {
    Long dp[][][][][];
    public long countBalanced(long low, long high) {
        return solve(high) - solve(low - 1);
    }

    public long solve(long num){
        dp = new Long[17][2][2][150][150];
        return helper(String.valueOf(num), 0, 1, 1, 0, 0);
    }

    public long helper(String str, int idx, int tight, int startZero, int sumOdd, int sumEven){
        if(idx >= str.length()) return (sumOdd == sumEven)? 1 : 0;
        if(dp[idx][tight][startZero][sumOdd][sumEven] != null) return dp[idx][tight][startZero][sumOdd][sumEven];
        
        int lb = 0;
        int ub = (tight == 1)? (int) (str.charAt(idx) - '0') : 9;

        long count = 0L;
        for(int i = lb; i <= ub; i++){
            if(i == 0 && startZero == 1){
                count += helper(str, idx + 1, 0, startZero, sumOdd, sumEven);
            } else {
                count += helper(str, idx + 1, (tight == 1 && i == ub)? 1 : 0, 0, (idx % 2 != 0)? sumOdd + i : sumOdd, (idx % 2 == 0)? sumEven + i : sumEven);
            }
        }
        return dp[idx][tight][startZero][sumOdd][sumEven] = count;
    }
}

File: solutions/4095-number-of-balanced-integers-in-a-range/number-of-balanced-integers-in-a-range.swift
final class Solution {
    func countBalanced(_ low: Int, _ high: Int) -> Int {
        func solve(_ x: Int) -> Int {
            if x < 0 { return 0 }
            let digits = Array(String(x)).map { Int(String($0))! }
            let n = digits.count
            let offset = 9 * n
            let size = offset * 2 + 1

            var memo = Array(repeating: Array(repeating: Array(repeating: Int.min, count: size), count: 2), count: n + 1)

            func dfs(_ idx: Int, _ tight: Int, _ diff: Int) -> Int {
                if idx == n { return diff == 0 ? 1 : 0 }
                let key = diff + offset
                if memo[idx][tight][key] != Int.min { return memo[idx][tight][key] }

                let sign = (idx & 1) == 0 ? 1 : -1
                let ub = tight == 1 ? digits[idx] : 9
                var res = 0

                for d in 0...ub {
                    let nt = (tight == 1 && d == ub) ? 1 : 0
                    res += dfs(idx + 1, nt, diff + sign * d)
                }

                memo[idx][tight][key] = res
                return res
            }

            return dfs(0, 1, 0)
        }

        return solve(high) - solve(low - 1)
    }
}